File: headless/src/main/org/nlogo/prim/_linkbreedsingular.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim;

import org.nlogo.agent.AgentSet;
import org.nlogo.agent.Link;
import org.nlogo.api.LogoException;
import org.nlogo.api.Nobody$;
import org.nlogo.api.Syntax;
import org.nlogo.nvm.Context;
import org.nlogo.nvm.Reporter;

public final strictfp class _linkbreedsingular
    extends Reporter {
  private final String breedName;

  public _linkbreedsingular(String breedName) {
    this.breedName = breedName;
  }

  @Override
  public Object report(final Context context) throws LogoException {
    AgentSet breed = world.getLinkBreed(breedName);
    Link link = world.getLink(argEvalDouble(context, 0),
        argEvalDouble(context, 1), breed);
    if (link == null) {
      return Nobody$.MODULE$;
    }
    return link;
  }

  @Override
  public String toString() {
    return super.toString() + ":" + breedName;
  }

  @Override
  public Syntax syntax() {
    return Syntax.reporterSyntax
        (new int[]{Syntax.NumberType(), Syntax.NumberType()},
            Syntax.LinkType() | Syntax.NobodyType());
  }
}

File: netlogo-gui/src/main/prim/_linkbreedsingular.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim

import org.nlogo.agent.{ AgentSet, Link }
import org.nlogo.api.{ LogoException, Syntax }
import org.nlogo.core.Nobody
import org.nlogo.nvm.{ Context, Reporter }

class _linkbreedsingular(breedName: String) extends Reporter {
  override def syntax =
    Syntax.reporterSyntax(
      Array[Int](Syntax.NumberType, Syntax.NumberType),
      Syntax.LinkType | Syntax.NobodyType)

  override def toString: String = s"${super.toString}:$breedName"

  override def report(context: Context): AnyRef = {
    val breed = world.getLinkBreed(breedName)
    val link = world.getLink(argEvalDouble(context, 0), argEvalDouble(context, 1), breed)
    if (link == null) Nobody else link
  }
}
